File: src/structure/namespace/collectNamespaceInfo.ts
import ts, {SourceFile, Node} from "typescript";
import {Configuration} from "../../configuration/configuration.js";
import {traverse} from "../../utils/traverse.js";
import {createNamespaceInfoItem, NamespaceInfoItem} from "./createNamespaceInfoItem.js";
import {InputStructureItem} from "../structure.js";
import {ImportInfo} from "../import/collectImportInfo.js";

interface InputNamespaceInfoItem extends NamespaceInfoItem, InputStructureItem {
}

export type NamespaceInfo = InputNamespaceInfoItem[]

export function collectNamespaceInfo(
    sourceFiles: SourceFile[],
    importInfo: ImportInfo,
    configuration: Configuration,
): NamespaceInfo {
    const result: NamespaceInfo = []

    sourceFiles.forEach(sourceFile => {
        traverse(sourceFile, node => {
            if (ts.isModuleDeclaration(node)) {
                const imports = importInfo.get(node) ?? []

                const item = createNamespaceInfoItem(
                    node,
                    sourceFile.fileName,
                    imports,
                    configuration,
                )

                let nodes: Node[] = []

                if (node.body && ts.isModuleBlock(node.body)) {
                    nodes = Array.from(node.body.statements)
                }

                result.push({
                    ...item,
                    nodes,
                    meta: {
                        type: "Namespace",
                        name: item.name
                    }
                })
            }
        })
    })

    return result
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/structure/namespace/collectNamespaceInfo.kt
package io.github.sgrishchenko.karakum.structure.namespace

import io.github.sgrishchenko.karakum.configuration.Configuration
import io.github.sgrishchenko.karakum.structure.InputStructureItem
import io.github.sgrishchenko.karakum.structure.InputStructureItemMeta
import io.github.sgrishchenko.karakum.structure.import.ImportInfo
import io.github.sgrishchenko.karakum.util.traverse
import js.array.ReadonlyArray
import js.objects.JsPlainObject
import js.objects.recordOf
import typescript.*

@JsPlainObject
external interface InputNamespaceInfoItem : NamespaceInfoItem, InputStructureItem

typealias NamespaceInfo = ReadonlyArray<InputNamespaceInfoItem>

fun collectNamespaceInfo(
    sourceFiles: ReadonlyArray<SourceFile>,
    importInfo: ImportInfo,
    configuration: Configuration,
): NamespaceInfo {
    val result = mutableListOf<InputNamespaceInfoItem>()

    sourceFiles.forEach { sourceFile ->
        traverse(sourceFile) { node ->
            if (isModuleDeclaration(node)) {
                val imports = importInfo[node] ?: emptyArray()

                val item = createNamespaceInfoItem(
                    node,
                    sourceFile.fileName,
                    imports,
                    configuration,
                )

                var nodes: ReadonlyArray<Node> = emptyArray()

                val body = node.body

                if (body != null && isModuleBlock(body)) {
                    nodes = body.statements.asArray()
                }

                // TODO: create ticket for JsPlainObject
                result += recordOf(
                    "fileName" to item.fileName,
                    "package" to item.`package`,
                    "moduleName" to item.moduleName,
                    "qualifier" to item.qualifier,
                    "hasRuntime" to item.hasRuntime,
                    "imports" to item.imports,

                    "name" to item.name,
                    "strategy" to item.strategy,

                    "nodes" to nodes,
                    "meta" to InputStructureItemMeta(
                        type = "Namespace",
                        name = item.name,
                    ),
                ).unsafeCast<InputNamespaceInfoItem>()
            }
        }
    }

    return result.toTypedArray()
}
